src/linear_algebra/vandermonde.lean

```diff
@@ -0,0 +1,177 @@
+  vandermonde v =
+    fin.cons (λ j, v 0 ^ (j : ℕ))
+      (λ i, fin.cons 1 (λ j, v i.succ * vandermonde (fin.tail v) i j)) :=
+begin
+  conv_lhs { rw [← fin.cons_self_tail v, vandermonde_cons] },
+  simp only [fin.tail]
+end
+
+@[simp]
+lemma fin.univ_filter_zero_lt {n : ℕ} :
+  (finset.univ : finset (fin n.succ)).filter (λ i, 0 < i) =
+    finset.univ.map ⟨fin.succ, fin.succ_injective _⟩ :=
+begin
+  ext i,
+  simp only [finset.mem_filter, finset.mem_map, finset.mem_univ, true_and,
+    function.embedding.coe_fn_mk, exists_true_left],
+  split,
+  { refine fin.cases _ _ i,
+    { rintro ⟨⟨⟩⟩ },
+    { intros i _, exact ⟨i, finset.mem_univ _, rfl⟩ } },
+  { rintro ⟨i, _, rfl⟩,
+    exact fin.succ_pos _ },
+end
+
+@[simp]
+lemma fin.univ_filter_succ_lt {n : ℕ} (j : fin n) :
+  (finset.univ : finset (fin n.succ)).filter (λ i, j.succ < i) =
+    (finset.univ.filter (λ i, j < i)).map ⟨fin.succ, fin.succ_injective _⟩ :=
+begin
+  ext i,
+  simp only [finset.mem_filter, finset.mem_map, finset.mem_univ, true_and,
+      function.embedding.coe_fn_mk, exists_true_left],
+  split,
+  { refine fin.cases _ _ i,
+    { rintro ⟨⟨⟩⟩ },
+    { intros i hi,
+      exact ⟨i, finset.mem_filter.mpr ⟨finset.mem_univ _, fin.succ_lt_succ_iff.mp hi⟩, rfl⟩ } },
+  { rintro ⟨i, hi, rfl⟩,
+    exact fin.succ_lt_succ_iff.mpr (finset.mem_filter.mp hi).2 },
+end
+
+-- `to_additive` version of `prod_filter_zero_lt, can't be autogenerated because
+-- of the `0` in the statement
+lemma sum_filter_zero_lt {M : Type*} [add_comm_monoid M] {n : ℕ} {v : fin n.succ → M} :
+  ∑ i in finset.univ.filter (λ (i : fin n.succ), 0 < i), v i = ∑ (j : fin n), v j.succ :=
+by rw [fin.univ_filter_zero_lt, finset.sum_map, function.embedding.coe_fn_mk]
+
+@[to_additive]
+lemma prod_filter_zero_lt {M : Type*} [comm_monoid M] {n : ℕ} {v : fin n.succ → M} :
+  ∏ i in finset.univ.filter (λ (i : fin n.succ), 0 < i), v i = ∏ (j : fin n), v j.succ :=
+by rw [fin.univ_filter_zero_lt, finset.prod_map, function.embedding.coe_fn_mk]
+
+@[to_additive]
+lemma prod_filter_succ_lt {M : Type*} [comm_monoid M] {n : ℕ} (j : fin n) (v : fin n.succ → M) :
+  ∏ i in finset.univ.filter (λ i, j.succ < i), v i =
+    ∏ j in finset.univ.filter (λ i, j < i), v j.succ :=
+by rw [fin.univ_filter_succ_lt, finset.prod_map, function.embedding.coe_fn_mk]
+
+/-- The "remarkable product" giving the differnece of two powers. -/
+lemma sub_mul_sum_pow (a b : R) (n : ℕ) :
+  (a - b) * ∑ i in finset.range (n + 1), a ^ i * b ^ (n - i) = a ^ (n + 1) - b ^ (n + 1) :=
+begin
+  induction n with n ih,
+  { simp },
+  calc (a - b) * ∑ (i : ℕ) in finset.range (n + 2), a ^ i * b ^ (n + 1 - i)
+      = a * ((a - b) * (∑ (x : ℕ) in finset.range (n + 1), a ^ x * b ^ (n - x))) +
+        (a - b) * (b * b ^ n) :
+    by simp [finset.sum_range_succ', pow_succ, mul_assoc, ← finset.mul_sum, mul_add, mul_left_comm]
+  ... = a * (a ^ (n + 1) - b ^ (n + 1)) + (a - b) * (b * b ^ n) : by rw ih
+  ... = a ^ (n + 2) - b ^ (n + 2) : by ring_exp
+end
```

Comment:
Oops, forgot to put these results in the correct files. Marking as WIP until I can fix that...